src/Nerdbank.MessagePack/Converters/ObjectArrayConverter`1.cs

```diff
@@ -36,159 +50,173 @@ internal class ObjectArrayConverter<T>(
 	/// <inheritdoc/>
 	public override T? Read(ref MessagePackReader reader, SerializationContext context)
 	{
-		if (reader.TryReadNil())
-		{
-			return default;
-		}
-
-		if (constructor is null)
+		try
 		{
-			throw new NotSupportedException($"The {typeof(T).FullName} type cannot be deserialized.");
-		}
+			if (reader.TryReadNil())
+			{
+				return default;
+			}
 
-		context.DepthStep();
-		T value = constructor();
-		IMessagePackSerializationCallbacks? callbacks = value as IMessagePackSerializationCallbacks;
-		callbacks?.OnBeforeDeserialize();
+			if (constructor is null)
+			{
+				throw new NotSupportedException($"The {typeof(T).FullName} type cannot be deserialized.");
+			}
 
-		UnusedDataPacket.Array? unused = null;
+			context.DepthStep();
+			T value = constructor();
+			IMessagePackSerializationCallbacks? callbacks = value as IMessagePackSerializationCallbacks;
+			callbacks?.OnBeforeDeserialize();
 
-		if (!typeof(T).IsValueType)
-		{
-			context.ReportObjectConstructed(value);
-		}
+			UnusedDataPacket.Array? unused = null;
 
-		if (reader.NextMessagePackType == MessagePackType.Map)
-		{
-			PropertyCollisionDetection collisionDetection = new(propertyShapes);
+			if (!typeof(T).IsValueType)
+			{
+				context.ReportObjectConstructed(value);
+			}
 
-			// The indexes we have are the keys in the map rather than indexes into the array.
-			int count = reader.ReadMapHeader();
-			for (int i = 0; i < count; i++)
+			if (reader.NextMessagePackType == MessagePackType.Map)
 			{
-				int index = reader.ReadInt32();
-				if (properties.Length > index && properties.Span[index] is { MsgPackReaders: var (deserialize, _), Shape.Position: int propertyShapePosition })
-				{
-					collisionDetection.MarkAsRead(propertyShapePosition);
-					deserialize(ref value, ref reader, context);
-				}
-				else if (unusedDataProperty?.Setter is not null)
-				{
-					unused ??= new();
-					unused.Add(index, reader.ReadRaw(context));
-				}
-				else
+				PropertyCollisionDetection collisionDetection = new(propertyShapes);
+
+				// The indexes we have are the keys in the map rather than indexes into the array.
+				int count = reader.ReadMapHeader();
+				for (int i = 0; i < count; i++)
```

Comment:
The indexer `i` should be defined outside the `try` block as an `int? i = null`, and the `catch` block should include the index in the exception message if it is non-null. 
Review your other changes and make similar adjustments for all array or map based serialization and deserialization.
Such an indexer (or property name property for the maps) should be reset to `null` at the conclusion of the loop so that exceptions thrown after the loop are not blamed on the last element.